service/history/shard/controller_impl.go

```diff
@@ -446,12 +464,63 @@ func (c *ControllerImpl) acquireShards(ctx context.Context) {
 	_ = sem.Acquire(ctx, concurrency)
 
 	c.RLock()
+
 	numOfOwnedShards := len(c.historyShards)
+	// Readiness check: We should set initialShardsAcquired when:
+	// 1. It's not already set.
+	// 2. We should own at least one shard (i.e. not before we join membership).
+	// 3. We have ownership of all the shards we're supposed to own.
+	if !c.initialShardsAcquired.Ready() && numOfOwnedShards > 0 {
+		go func() {
+			defer readinessCancel()
+			c.checkShardReadiness(readinessCtx, expmaps.Values(c.historyShards))
+		}()
+	} else {
+		readinessCancel()
+	}
+
 	c.RUnlock()
 	metrics.NumShardsGauge.With(c.taggedMetricsHandler).Record(float64(numOfOwnedShards))
 	c.publishShardCountUpdate(numOfOwnedShards)
 }
 
+func (c *ControllerImpl) checkShardReadiness(
```

Comment:
A suggestion: this could have return value(s) that indicate whether all the shards are ready,  then all of the concurrency/cancelation logic could live in `acquireShards` instead of being spread between the 2 functions.

Reply:
After Joe's suggestion I think this is the case? Are you suggesting just moving the last few lines (with `c.initialShardsAcquired.SetIfNotReady`) into acquireShards? I guess that's okay but I don't think it's an improvement

Reply:
> Are you suggesting just moving the last few lines ...

yes, that's my suggestion. I think it's a little easier to understand the management of `c.initialShardsAcquired` if all of the modifications are in the same function. But up to you.

Reply:
ok, that's a good point